feat : token hub

matijamarjanovic opened the pull request:
### Description

Token hub is a registry for tokens created on gno.land. Since all the tokens are visible in each their own realms, as it is right now there is no easy way that a user can check which tokens belong to him by his address. Token hub bypasses this by concentrating each each token in it's own realm and allows operations over these tokens providing a easier way to communicate with the realms in which the tokens are originally made. 

### How it works
In the realm where token is being made, tokenhub needs to be imported and the newly made token has to be registered to the token hub by using RegisterToken function and passing the token as an argument. This is everything token hub needs to be able to operate.

### Contributor's checklist
- [x] Add GRC20 support for token registration and checking of balance
- [ ] Add support for the rest of GRC20 opperations such as transfers, minting, burning, etc. 
- [ ] Add support for different types of tokens such as GRC721 and GRC1155
- [ ] Consider security and privacy of users 
- [ ] Add tests

All feedback is welcome

Gno2D2 commented:
#### 🛠 PR Checks Summary
All **Automated Checks** passed. ✅

##### Manual Checks (for Reviewers):
- [ ] **IGNORE** the bot requirements for this PR (force green CI check)
- [ ] The pull request description provides enough details


<details><summary>Read More</summary>

🤖 This bot helps streamline PR reviews by verifying automated checks and providing guidance for contributors and reviewers.

##### ✅ Automated Checks (for Contributors):
 🟢 Maintainers must be able to edit this pull request ([more info](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork))


##### ☑️ Contributor Actions:
1. Fix any issues flagged by automated checks.
2. Follow the Contributor Checklist to ensure your PR is ready for review.
	- Add new tests, or document why they are unnecessary.
	- Provide clear examples/screenshots, if necessary.
	- Update documentation, if required.
	- Ensure no breaking changes, or include `BREAKING CHANGE` notes.
	- Link related issues/PRs, where applicable.

##### ☑️ Reviewer Actions:
1. Complete manual checks for the PR, including the guidelines and additional checks if applicable.

##### 📚 Resources:
- [Report a bug with the bot](https://github.com/gnolang/gno/issues/3238).
- [View the bot’s configuration file](https://github.com/gnolang/gno/tree/master/contribs/github-bot/internal/config/config.go).

<details><summary><b>Debug</b></summary><blockquote>
<details><summary><b>Automated Checks</b></summary><blockquote>

<details><summary>Maintainers must be able to edit this pull request (more info)</summary><blockquote>

### If
```
🟢 Condition met
└── 🟢 The pull request was created from a fork (head branch repo: matijamarjanovic/gno)

```
### Then
```
🟢 Requirement satisfied
└── 🟢 Maintainer can modify this pull request

```
</blockquote></details>

</blockquote></details>


<details><summary><b>Manual Checks</b></summary><blockquote>

<details><summary>**IGNORE** the bot requirements for this PR (force green CI check)</summary><blockquote>

### If
```
🟢 Condition met
└── 🟢 On every pull request

```
### Can be checked by

- Any user with comment edit permission

</blockquote></details>

<details><summary>The pull request description provides enough details</summary><blockquote>

### If
```
🟢 Condition met
└── 🟢 Not (🔴 Pull request author is a member of the team: core-contributors)

```
### Can be checked by
 - team core-contributors

</blockquote></details>

</blockquote></details>

</blockquote></details>

</details>

leohhhn commented:
Check this out, it's a similar idea: https://github.com/gnolang/gno/blob/master/examples/gno.land/r/demo/grc20reg/grc20reg.gno

matijamarjanovic commented:
> Check this out, it's a similar idea
> 
> 
> 
> 

Thanks for pointing out, I will consider implementation for different token types

leohhhn commented on examples/gno.land/r/demo/grc20reg/grc20reg.gno:
```diff
@@ -73,4 +73,31 @@ func Render(path string) string {
 	}
 }
 
+func Iterate(callback func(key string, token grc20.TokenGetter) bool) {
+	registry.Iterate("", "", func(key string, tokenI interface{}) bool {
+		tokenGetter := tokenI.(grc20.TokenGetter)
+		return callback(key, tokenGetter)
+	})
+}
+
+// iterate through n items starting from the startKey (pagination)
+func IterateN(startKey string, n int, callback func(key string, token grc20.TokenGetter) bool) string {
+	var lastKey string
+	count := 0
+
+	registry.Iterate(startKey, "", func(key string, tokenI interface{}) bool {
+		if count >= n {
+			return true
+		}
+
+		tokenGetter := tokenI.(grc20.TokenGetter)
+		lastKey = key
+		count++
+
+		return callback(key, tokenGetter)
+	})
+
+	return lastKey
+}
+
```
Don't do this. Make a readonly tree getter instead (check avl folder), and use the `avl/pager` library in your smart contract.

matijamarjanovic replied:
Thanks for the advice, the code looks much better now and i can imagine the performance should be better as well. See changes 59586e50ccccfd7ef9056d294b7d249d530b6c0e

leohhhn commented on examples/gno.land/r/matijamarjanovic/tokenhub/tokenhub.gno:
```diff
@@ -0,0 +1,534 @@
+	"gno.land/r/demo/users"
+)
+
+type GRC1155TokenInfo struct {
+	Collection grc1155.MultiTokenGetter
+	TokenID    string
+}
```
I'm wondering if this is the right place for this. Is it specific to your hub, or maybe should it be something that exists in the 1155 package itself? Put some godoc as to why you need this.

matijamarjanovic replied:
Actually, I think grc1155 need a pretty big refactor as I was able to change other people's balances using this getter so I think adding this to it wouldn't make much difference. 
I think this tokenhub should keep this info for now as I expect having to change it in the future with the GRC1155 changes. Please correct me if this is a wrong approach

leohhhn commented on examples/gno.land/r/matijamarjanovic/tokenhub/tokenhub.gno:
```diff
@@ -0,0 +1,534 @@
+		}
+		sb.WriteString(md.BulletList(links))
+
+	case strings.HasPrefix(path, "token"): // grc20
```
If a string need a comment to explain it like this, maybe it's better to define it as a constant.

matijamarjanovic replied:
fixed in 43b4b5227cc72bb6781b1c290a043a3d8920bc17

leohhhn commented on examples/gno.land/r/matijamarjanovic/tokenhub/tokenhub.gno:
```diff
@@ -0,0 +1,534 @@
+}
+
+func getBalances(input string) string {
+
```
random newline

matijamarjanovic replied:
removed in 1c4b72a4336cd9afb2f95497e4e56126b1e6e0a9

leohhhn commented on examples/gno.land/r/matijamarjanovic/tokenhub/tokenhub.gno:
```diff
@@ -0,0 +1,534 @@
+	balances := avl.NewTree()
+	var sb strings.Builder
+
+	registeredNFTs.Iterate("", "", func(key string, value interface{}) bool {
+		nftGetter := value.(grc721.NFTGetter)
+		nft := nftGetter()
+		key_parts := strings.Split(key, ".")
+		owner, err := nft.OwnerOf(grc721.TokenID(key_parts[len(key_parts)-1]))
+		if err == nil && addr == owner { // show only the nfts owner owns
+			balances.Set("nft:"+key, 1)
+		}
+		return false
+	})
+
+	grc20reg.Iterate(func(key string, tokenGetter grc20.TokenGetter) bool {
+		token := tokenGetter()
+		balance := token.BalanceOf(addr)
+		balances.Set("token:"+key, balance)
+		return false
+	})
+
+	registeredMTs.Iterate("", "", func(key string, value interface{}) bool {
+		info := value.(GRC1155TokenInfo)
+		mt := info.Collection()
+		balance, err := mt.BalanceOf(addr, grc1155.TokenID(info.TokenID))
+		if err == nil {
+			balances.Set("mt:"+key, balance)
+		}
+		return false
+	})
```
This will be a pretty bad case when your hub gains popularity. You will most likely break gas limits.

Try to optimize this somehow, split it up, not sure exactly how. Iterating trees should be avoided wherever possible.

matijamarjanovic replied:
As we agreed in person, the only thing I could to to optimize this is iterate the lists separately instead of all at once. This was done in https://github.com/gnolang/gno/pull/3479/commits/9b9767f56b6c4b1eb9bde33ce893c9d4e9ae6c4d

leohhhn commented on examples/gno.land/r/matijamarjanovic/tokenhub/tokenhub.gno:
```diff
@@ -0,0 +1,534 @@
+	return sb.String()
+}
+
+func getAddressForUsername(addrOrName string) std.Address {
+	addr := std.Address(addrOrName)
+
+	if addr.IsValid() {
+		return addr
+	}
+	if user := users.GetUserByName(addrOrName); user != nil {
+		return user.Address
+	}
+
+	return ""
+}
```
```suggestion
func getAddressForUsername(addrOrName string) std.Address {
	addr := std.Address(addrOrName)
	if addr.IsValid() {
		return addr
	}
	
	if user := users.GetUserByName(addrOrName); user != nil {
		return user.Address
	}
	
	return ""
}
```

matijamarjanovic replied:
Is this just a suggestion to remove \n?

leohhhn reviewed (changes requested):
Good start but there's more to be worked on. Check out the comments I left below.

matijamarjanovic commented on examples/gno.land/r/matijamarjanovic/tokenhub/tokenhub.gno:
```diff
@@ -0,0 +1,534 @@
+	ErrNFTNotMetadata       = errors.New("NFT must implement IGRC721CollectionMetadata")
+
+	pageSize = 2
+	test     = ""
```
Fixed this in 8c097251df0ffc0d1dc7a2d9a6279796dcb72d1f

matijamarjanovic commented on examples/gno.land/r/matijamarjanovic/tokenhub/tokenhub.gno:
```diff
@@ -0,0 +1,534 @@
+	ErrNFTNotFound          = errors.New("NFT not found")
+	ErrMTAlreadyRegistered  = errors.New("Multi-token already registered")
+	ErrMTNotFound           = errors.New("Multi-token not found")
+	ErrMTInfoNotFound       = errors.New("Multi-token info not found")
```
noted. see 88b0a7ee5436de12426a91d52f22b215e0b7f223 for the fix

matijamarjanovic commented on examples/gno.land/r/matijamarjanovic/tokenhub/tokenhub.gno:
```diff
@@ -0,0 +1,534 @@
+
+// registering a collection of NFTs
+func RegisterNFT(nftGetter grc721.NFTGetter, collection string, tokenId string) error {
+
```
removed it in 1c4b72a4336cd9afb2f95497e4e56126b1e6e0a9

matijamarjanovic commented on examples/gno.land/r/matijamarjanovic/tokenhub/tokenhub.gno:
```diff
@@ -0,0 +1,534 @@
+	nftOwner, err := nft.OwnerOf(grc721.TokenID(tokenId))
+	if string(nftOwner) == "" || err != nil {
+		return ErrNFTtokIDNotExists
+	}
```
good point, fixed this in : 005e64b635dfcabdfb227aff84d41787a603a30d

matijamarjanovic commented on examples/gno.land/r/matijamarjanovic/tokenhub/tokenhub.gno:
```diff
@@ -0,0 +1,534 @@
+		slug = slugs[0]
+	}
+	grc20reg.Register(tokenGetter, slug)
+}
```
Now I realize this might be an overkill, but what I was going for is this: slugs should be either 0 length or 1, if it is more the rest is ignored. I changed this to just be one string in : 005e64b635dfcabdfb227aff84d41787a603a30d

matijamarjanovic commented on examples/gno.land/r/matijamarjanovic/tokenhub/tokenhub.gno:
```diff
@@ -0,0 +1,534 @@
+
+func GetMyBalances() string {
+	return getNonZeroBalances(string(std.PrevRealm().Addr()))
+}
```
Okay let me explain how I originally made the tokenhub. The idea was that only the owner is able to see all of his balances, so basically this would be used if someone wanted to keep some assets inside of some realm (the token.OwnerOf would return that realms address) - in this case user would use GetMyBalances inside of that mediator realm which would return balance of that realm. If you check out the 2 functions bellow labeled for `user use` it was meant to be used through that mediator contract but to get the users funds.

Either way, I realised tha approach was unnecessary in general as the balances of each token is public so it doesn't make sense to hide it here. GetMyBalances does the same as GetBalances(myAddress std.Address)

This problem was addressed here: 7ac568f7cbc18d36ce1fc7b726762d6dc14df913

matijamarjanovic commented on examples/gno.land/r/matijamarjanovic/tokenhub/tokenhub.gno:
```diff
@@ -0,0 +1,534 @@
+
+func GetMyBalancesOrigCaller() string {
+	return getNonZeroBalances(string(std.GetOrigCaller()))
+}
```
Thank you for advice on go conventions. I will also make sure to add godoc to all the functions. As for this case, I will delete these two functions for the reasons mentioned in previous response. Deletion in : 7ac568f7cbc18d36ce1fc7b726762d6dc14df913

matijamarjanovic commented on examples/gno.land/r/matijamarjanovic/tokenhub/tokenhub.gno:
```diff
@@ -0,0 +1,534 @@
+		panic("unknown multi-token: " + key)
+	}
+	return info
+}
```
Good question. I have changed this in 0a63572698f787a524948b1a7e869e9c4f721665 and 99b44ac30e964636bf3760f5aa1b31dad6f866b7

matijamarjanovic commented on examples/gno.land/r/matijamarjanovic/tokenhub/tokenhub.gno:
```diff
@@ -0,0 +1,534 @@
+	test     = ""
+)
+
+func RegisterToken(tokenGetter grc20.TokenGetter, slugs ...string) {
```
I have updated the whole realm's namings in 3f8bb518008f482ffea9852e9786b3fc6f0552e7. Hopefully it makes more sense now

matijamarjanovic commented on examples/gno.land/r/matijamarjanovic/tokenhub/tokenhub.gno:
```diff
@@ -0,0 +1,534 @@
+
+	rlmPath := std.PrevRealm().PkgPath()
+	k := ufmt.Sprintf("%s.%s", collection, tokenId)
+	key := fqname.Construct(rlmPath, k)
```
Okay this makes sense, it was originally meant to have the same approach as in grc20reg but it later needed changing so i just added concat. I removed fqname and used concat only in commit: b5a8206c11594eec8106f0c0542844d303f923eb

matijamarjanovic commented on examples/gno.land/r/matijamarjanovic/tokenhub/tokenhub.gno:
```diff
@@ -0,0 +1,534 @@
+	return nil
+}
+
+// for grc1155 is a little bit different, since the one collection can have multiple tokens, this is why don't use slugs but tokenIDs,
+// and the collection is saved as a part of the :value in the node
```
This was left as a reminder to me and more of a note to the reviewer, I will make sure to write proper godocs at the end of making req changes.

matijamarjanovic commented on examples/gno.land/r/matijamarjanovic/tokenhub/tokenhub.gno:
```diff
@@ -0,0 +1,534 @@
+func RegisterGRC1155Token(mtGetter grc1155.MultiTokenGetter, tokenID string) error {
+	rlmPath := std.PrevRealm().PkgPath()
+
+	key := fqname.Construct(rlmPath, tokenID)
```
Fixed in b5a8206c11594eec8106f0c0542844d303f923eb

matijamarjanovic commented on examples/gno.land/r/matijamarjanovic/tokenhub/tokenhub.gno:
```diff
@@ -0,0 +1,534 @@
+	return info
+}
+
+func GetAllNFTs() string {
```
This was done in : 3c5d8dc5174f58d17a79d0d2a801eb23bf07fc55

matijamarjanovic commented on examples/gno.land/r/matijamarjanovic/tokenhub/tokenhub.gno:
```diff
@@ -0,0 +1,534 @@
+
+func Render(path string) string {
+	var sb strings.Builder
+	sb.WriteString(test)
```
Fixed in 43b4b5227cc72bb6781b1c290a043a3d8920bc17

matijamarjanovic commented on examples/gno.land/r/matijamarjanovic/tokenhub/tokenhub.gno:
```diff
@@ -0,0 +1,534 @@
+	var sb strings.Builder
+	sb.WriteString(test)
+
+	switch {
```
Sure, implemented in 43b4b5227cc72bb6781b1c290a043a3d8920bc17

matijamarjanovic commented on examples/gno.land/r/matijamarjanovic/tokenhub/tokenhub.gno:
```diff
@@ -0,0 +1,534 @@
+		links := []string{
+			md.Link("GRC20 Tokens", "http://localhost:8888/r/matijamarjanovic/tokenhub:token_1"),
+			md.Link("GRC721 NFTs", "http://localhost:8888/r/matijamarjanovic/tokenhub:nft_1"),
+			md.Link("GRC1155 Multi-Tokens", "http://localhost:8888/r/matijamarjanovic/tokenhub:mt_1"),
```
So this is the homepage and these are the links to page 1 of the requested type, these are actually filters since I wanted to present only 1 token type in 1 page. When you go on these pages you would have a button at the bottom saying return to homepage that returns you to the home:
![image](https://github.com/user-attachments/assets/2958ae7e-0812-4d55-9319-e897aef55365)

For exapmple this is first page of pageSize = 2 for tokens:
![image](https://github.com/user-attachments/assets/00ab6350-8fe6-422a-9f56-aa2cb5932f0b)

thehowl commented:
Can you fix tests?

matijamarjanovic commented:
> Can you fix tests?

Could it be the coverage? I might have missed something
![image](https://github.com/user-attachments/assets/c9b1367a-7662-4f56-b8ce-cb1e3da8e72c)

leohhhn commented on examples/gno.land/r/demo/tokenhub/render.gno:
```diff
@@ -0,0 +1,166 @@
+	case path == "":
+		out = RenderHome()
+
+	case strings.HasPrefix(path, token):
+		out = RenderToken(path)
+
+	case strings.HasPrefix(path, nft):
+		out = RenderNFT(path)
+
+	case strings.HasPrefix(path, mt):
+		out = RenderMT(path)
```
Define these consts in this file, since this is where they're used, for readability. also use a const block an not individual declarations.

leohhhn commented on examples/gno.land/r/demo/tokenhub/getters.gno:
```diff
@@ -0,0 +1,208 @@
+	grc20reg.GetRegistry().Iterate("", "", func(key string, value interface{}) bool {
+		tokenGetter := value.(grc20.TokenGetter)
+		token := tokenGetter()
+		balance := token.BalanceOf(addr)
+		if nonZero && balance > 0 {
+			out += ufmt.Sprintf("Token:%s:%d,", key, balance)
+		} else if !nonZero {
+			out += ufmt.Sprintf("Token:%s:%d,", key, balance)
+		}
```
This can be simplified, very neatly

```go
balance := token.BalanceOf(addr)
if !nonZero || balance > 0 {
	out += ufmt.Sprintf("Token:%s:%d,", key, balance)
}
```

matijamarjanovic replied:
Right, got it in edd676b9e78661d41211205e34269f0e223e6863

leohhhn commented on examples/gno.land/r/demo/tokenhub/getters.gno:
```diff
@@ -0,0 +1,208 @@
+		mt := info.Collection()
+		balance, err := mt.BalanceOf(addr, grc1155.TokenID(info.TokenID))
+		if err == nil {
+			if nonZero && balance > 0 {
+				out += ufmt.Sprintf("MultiToken:%s:%d,", key, balance)
+			} else if !nonZero {
+				out += ufmt.Sprintf("MultiToken:%s:%d,", key, balance)
+			}
+		}
```
Same as above

matijamarjanovic replied:
got it in edd676b9e78661d41211205e34269f0e223e6863

leohhhn reviewed (commented):
Hey, decent job but still some more work to be done. Left some comments - also, please move this to your namespace :)

matijamarjanovic commented on examples/gno.land/r/demo/tokenhub/render.gno:
```diff
@@ -0,0 +1,166 @@
+		out = RenderNFT(path)
+
+	case strings.HasPrefix(path, mt):
+		out = RenderMT(path)
```
Nope, there is not. Fixed this in 13d327c6b6aba9d1f480f297be3602ba97a53a33

matijamarjanovic commented on examples/gno.land/r/demo/tokenhub/render.gno:
```diff
@@ -0,0 +1,166 @@
+	default:
+		out = md.H1("404 Not Found")
+		out += md.Paragraph("The requested page does not exist.")
+		out += md.Link("Back to home", "http://localhost:8888/r/demo/tokenhub")
```
forgot to change it. fixed links in 2a2a02910b4e953f52a313d51cff0dcaeb7377d7

matijamarjanovic commented on examples/gno.land/r/demo/tokenhub/render.gno:
```diff
@@ -0,0 +1,166 @@
+	if path != "" {
+		out += "\n"
+		out += md.Link("Back to home", "http://localhost:8888/r/demo/tokenhub")
+	}
```
Links fixed in 2a2a02910b4e953f52a313d51cff0dcaeb7377d7. Second part was done in b9580c804b60e2f1035390149074c3dfdbea078c.

matijamarjanovic commented on examples/gno.land/r/demo/tokenhub/render.gno:
```diff
@@ -0,0 +1,166 @@
+		md.Link("GRC20 Tokens", "http://localhost:8888/r/demo/tokenhub:tokens"),
+		md.Link("GRC721 NFTs", "http://localhost:8888/r/demo/tokenhub:nfts"),
+		md.Link("GRC1155 Multi-Tokens", "http://localhost:8888/r/demo/tokenhub:mts"),
+	}
```
2a2a02910b4e953f52a313d51cff0dcaeb7377d7

matijamarjanovic commented on examples/gno.land/r/demo/tokenhub/render.gno:
```diff
@@ -0,0 +1,166 @@
+	return out
+}
+
+func RenderHome() string {
```
nope. fixed in 13d327c6b6aba9d1f480f297be3602ba97a53a33

matijamarjanovic commented on examples/gno.land/r/demo/tokenhub/render.gno:
```diff
@@ -0,0 +1,166 @@
+	return out
+}
+
+func RenderToken(path string) string {
```
13d327c6b6aba9d1f480f297be3602ba97a53a33

matijamarjanovic commented on examples/gno.land/r/demo/tokenhub/render.gno:
```diff
@@ -0,0 +1,166 @@
+	return out
+}
+
+func RenderNFT(path string) string {
```
hahahaha 13d327c6b6aba9d1f480f297be3602ba97a53a33

matijamarjanovic commented on examples/gno.land/r/demo/tokenhub/render.gno:
```diff
@@ -0,0 +1,166 @@
+
+	page := nftPager.MustGetPageByPath(path)
+
+	for _, item := range page.Items {
```
addressed in 9ee19b1124300debd7b97211125e78da0769200e

matijamarjanovic commented:
> Hey, decent job but still some more work to be done. Left some comments - also, please move this to your namespace :)

Thanks, addressed everything. Moved it back to my namespace in 5fa00c8981274982bd3fdb55916707557325142a

jefft0 reviewed (approved):
Approving to remove the `review/triage-pending` label, and handing off to @leohhhn to manage the review process.

leohhhn commented:
Hey @matijamarjanovic, I started reviewing but you'll need to update the branch with the newest changes (ie the user registry system)

leohhhn commented on examples/gno.land/r/matijamarjanovic/tokenhub/render.gno:
```diff
@@ -0,0 +1,170 @@
+			md.InlineCode(item.Key)))
+	}
+
+	if len(nftItems) > 0 {
+		out += md.BulletList(nftItems)
+		out += "\n"
+
+		if picker := page.Picker(); picker != "" {
+			out += md.HorizontalRule()
+			out += md.Paragraph(picker)
+		} else {
+			out += md.HorizontalRule()
+		}
+	} else {
+		out += md.Italic("No NFTs registered yet")
+		out += "\n"
+	}
```
This is repeated code (x3), I think you can extract it to a separate function as you're dealing with strings

leohhhn commented on examples/gno.land/r/matijamarjanovic/tokenhub/render.gno:
```diff
@@ -0,0 +1,170 @@
+}
+func renderFooter() string {
+	out := "\n"
+	out += md.Link("Back to home", "http://localhost:8888/r/matijamarjanovic/tokenhub")
```
This link won't work :)

matijamarjanovic replied:
fixed in d2a32396655e7ae3a9d1de115c3b431a5b602af5

leohhhn reviewed (approved):
Looks good more or less, let's see it in action. I've left a few comments that should be resolved, and also master changes should be fixed.

moul commented on examples/gno.land/r/demo/grc20reg/grc20reg.gno:
```diff
@@ -22,6 +23,22 @@ func Register(tokenGetter grc20.TokenGetter, slug string) {
 	)
 }
 
+func RegisterWithTokenhub(tokenGetter grc20.TokenGetter, key string) {
+	prevRealmPath := std.PreviousRealm().PkgPath()
+	if prevRealmPath != "gno.land/r/matijamarjanovic/tokenhub" {
+		return
+	}
+
+	registry.Set(key, tokenGetter)
+
+	rlmPath, slug := fqname.Parse(key)
+	std.Emit(
+		registerEvent,
+		"pkgpath", rlmPath,
+		"slug", slug,
+	)
+}
```
We cannot have this in grc20reg.

The only way to extend grc20reg in this manner is by making it managed by a DAO or by having a list of whitelisted controllers that the DAO will manage. However, there is no way for this contract to check for a hardcoded dependency like this.

matijamarjanovic replied:
I thought this might represent a problem, for some time even I had a "Dillema" section in the description but removed it since I wanted to continue working on https://github.com/gnolang/gno/pull/3793 which depends on this. This seemed to be the best approach at the time as the alternatives were either to place all the content of this package in the same package as `grc20reg.gno` or to have a whole separate registry apart from that existing one. Of course, none of those options seem viable still

I am going to put some more thought into resolving this problem and I'm open to any suggestions/advice from anyone

Thank you for the feedback 🙏

moul commented on examples/gno.land/r/matijamarjanovic/tokenhub/getters.gno:
```diff
@@ -0,0 +1,216 @@
+	"gno.land/p/demo/grc/grc721"
+	"gno.land/p/demo/ufmt"
+	"gno.land/r/demo/grc20reg"
+	"gno.land/r/demo/users"
```
deprecated

matijamarjanovic replied:
updated with the new version: `/r/sys/users`, in 6e0a23305cb97daf8e0de5078a4cb2a96cb199cd

moul commented on examples/gno.land/p/demo/grc/grc721/basic_nft.gno:
```diff
@@ -395,3 +395,10 @@ func (s *basicNFT) RenderHome() (str string) {
 	return
 }
+
+// Then add the Getter method to your NFT types
```
????

matijamarjanovic replied:
Left the comment unintentionally, removed in d2a32396655e7ae3a9d1de115c3b431a5b602af5

moul reviewed (changes requested):
blocking because of the changes on r/demo/grc20reg

matijamarjanovic commented on examples/gno.land/r/matijamarjanovic/tokenhub/render.gno:
```diff
@@ -0,0 +1,170 @@
+	default:
+		out = md.H1("404 Not Found")
+		out += md.Paragraph("The requested page does not exist.")
+		out += "[Back to home](/r/matijamarjanovic/tokenhub)"
```
Makes sense, implemented in 6e0a23305cb97daf8e0de5078a4cb2a96cb199cd

matijamarjanovic commented on examples/gno.land/r/matijamarjanovic/tokenhub/render.gno:
```diff
@@ -0,0 +1,170 @@
+	} else {
+		out += md.Italic("No tokens registered yet")
+		out += "\n"
+	}
```
simplified it in d2a32396655e7ae3a9d1de115c3b431a5b602af5 in a new func called `renderItemsList()`

matijamarjanovic commented:
I have removed the hardcoded dependency in grc20reg by simply concating the original realm path to the slug. Although the key registered through tokenhub is going to be twice as long now I think this might be the best approach. Please let me know if this causes any further inconvenience

e.g.:
![image](https://github.com/user-attachments/assets/07fd6fd1-d792-4004-86c8-4015aeb04db0)
